### DISCUSSION TITLE
Linux bundle not portable: numpy -> glibc

### DISCUSSION THREAD

**Discussion by clstaudt** [pyinstaller/pyinstaller#7148] [Help]
I am trying trying to package a Python app that uses numpy for Linux. While the executable runs on the system on which it was built, it does not seem portable to other Linux systems. Could you please advise on where the error comes from and how to resolve it?

https://github.com/tuttle-dev/tuttle/issues/56

```
> ./Tuttle-linux-x86

ImportError: Unable to import required dependencies:
numpy: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.10 from "/home/tobias/Downloads/Tuttle-linux-x86"
  * The NumPy version is: "1.23.3"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.35' not found (required by /tmp/_MEIt9SVU6/././libgcc_s.so.1)

[670635] Failed to execute script 'Tuttle' due to unhandled exception!
```

**Comment by rokm** [comment]
> Original error was: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.35' not found (required by /tmp/_MEIt9SVU6/././libgcc_s.so.1)

The `libgcc_s.so.1`  that was collected from your build system was linked against more recent version of `glibc` that is available on your target system, hence the symbol error. Generally, you should build your program on a system with the lowest version of glibc that you want to support (i.e., the program will likely not run on systems with older version of glibc).

You might also try removing `libgcc_s.so.1` from your bundle (it is one of those libraries that we should re-consider collecting, as per #6993), although the only way to achieve that in `onefile` builds is to edit your .spec file and filter `Analysis.binaries` list...

**Reply by clstaudt** [reply]
Thanks for the explanation @rokm . What should I do if I do not want to have an opinion about which version of `glibc` to support? Since my project does not depend directly on a version of `glibc`, I'd rather not introduce new machinery to pin versions of system-level libraries.  Would removing `libgcc_s.so.1` help to evade the problem?

**Reply by rokm** [reply]
> Thanks for the explanation @rokm . What should I do if I do not want to have an opinion about which version of `glibc` to support? I'd rather not introduce new machinery to pin versions of system-level libraries. 

Unless your program is really simple in terms of external dependencies (i.e., the system libraries that used python extensions are linked against), you likely won't be able to create a "universal" linux binary, if that's what you're hoping for. 

Because as soon as PyInstaller collects a shared library from your build system (due to dependency from some other collected binary, like a binary python extension), your program requires `glibc` that is of the same or greater version than the one found on your build system. 

So you can not have an opinion about supported `glibc` versions, but you effectively made a choice as soon as you picked up your build system.

> Would removing `libgcc_s.so.1` help to evade the problem?

It might (assuming the compatible version of library is available on the target system). But there might be other shared libraries, which you might or might not be able to remove, depending on what you expect to find on the target system.

That's why we usually recommend you build your program in an older version of debian docker container. But for programs with more complex dependencies, that might cause other problems due to forward-incompatible ABI changes, or even distribution-patched behavior (e.g., #4985). It is all very much situation-dependent.

All in all, distributing binary software on linux is not a very fun endeavor...

**Reply by clstaudt** [reply]
@rokm Thanks, that wasn't what I hoped for but a good explanation of what is going on. 

> That's why we usually recommend you build your program in an older version of debian docker container. 

Do you happen to know how I can get such a container via GitHub actions?  Currently my build action starts with:
```
jobs:
  build:
    runs-on: ${{ matrix.os }}
    strategy:
      fail-fast: false
      matrix:
        os: ['windows-latest', 'ubuntu-latest', 'macos-latest']
```

**Reply by rokm** [reply]
You can run the container from within the GitHub workflow; for example, we run our test suite in Alpine container ([here](https://github.com/pyinstaller/pyinstaller/blob/17f7209eafb7fa4234d98b7b21e55ca93d90c538/.github/workflows/ci.yml#L146-L171)) as part of the CI. You'll need to share a directory and use it as working directory within the container (see for example https://github.com/pyinstaller/pyinstaller/issues/7142#issuecomment-1272579831). There's also seem to be GitHub actions available that integrate docker + pyinstaller, but I have no idea if they are up-to-date nor if they work.
